File: engine/src/main/java/com/mjaseem/trick/engine/Card.java
package com.mjaseem.trick.engine;

// Card class
public record Card(Records.Suit suit, int rank) {

    public boolean beats(Card other, Records.Suit leadSuit, Records.Suit trumpSuit) {
        if (this.suit == trumpSuit && other.suit() != trumpSuit) {
            return true;
        }
        if (this.suit != trumpSuit && other.suit == trumpSuit) {
            return false;
        }
        if (this.suit == other.suit()) {
            return this.rank > other.rank();
        }
        if (other.suit() == leadSuit) {
            return false;
        }
        return this.suit == leadSuit;
    }

    @Override
    public String toString() {
        String suitSymbol = switch (suit) {
            case HEARTS -> "♥";
            case CLUBS -> "♣";
            case DIAMONDS -> "♦";
            case SPADES -> "♠";
        };
        return rank + suitSymbol;
    }
}

File: rl/engine/card.py
from enum import Enum


class Suit(Enum):
    HEARTS = "♥"
    CLUBS = "♣"
    DIAMONDS = "♦"
    SPADES = "♠"


class Card:
    def __init__(self, suit: Suit, rank: int) -> None:
        self.suit = suit
        self.rank = rank

    def beats(self, other: "Card", lead_suit: Suit, trump_suit: Suit) -> bool:
        if self.suit == trump_suit and other.suit != trump_suit:
            return True
        if self.suit != trump_suit and other.suit == trump_suit:
            return False
        if self.suit == other.suit:
            return self.rank > other.rank
        if other.suit == lead_suit:
            return False
        return self.suit == lead_suit

    def __str__(self) -> str:
        return f"{self.rank}{self.suit.value}"

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Card):
            return False
        return self.suit == other.suit and self.rank == other.rank